python/cudf/cudf/core/column/numerical.py

```diff
@@ -109,13 +99,28 @@ def binary_operator(
                 msg = "{!r} operator not supported between {} and {}"
                 raise TypeError(msg.format(binop, type(self), type(rhs)))
             out_dtype = np.result_type(self.dtype, rhs.dtype)
-            if binop in ["mod", "floordiv"]:
-                tmp = self if reflect else rhs
-                if (tmp.dtype in int_dtypes) and (
-                    (np.isscalar(tmp) and (0 == tmp))
-                    or ((isinstance(tmp, NumericalColumn)) and (0.0 in tmp))
-                ):
-                    out_dtype = np.dtype("float64")
+
+            # Special cases for mod and floordiv
+            if binop in {"mod", "floordiv"}:
+                denominator = self if reflect else rhs  # denominator
+                numerator = rhs if reflect else self  # numerator
+                # check for zeroes in the denominator
+                if denominator.dtype.kind in "bui":
+                    # Denominator either zero, or zero is contained therein
+                    if (np.isscalar(denominator) and denominator == 0) or (
+                        isinstance(denominator, NumericalColumn)
+                        and 0 in denominator
```

Comment:
This is just really expensive (because it forces synchronization) to always do and differs in behavior from Numpy / Pandas. Do we have any other options here?

Reply:
Pandas a numpy give zero for this, numpy raises a RuntimeWarning and then also returns zero. I could see us raising a warning as well as @cwharris suggested, but ultimately we can't know if we should raise the warning or not without knowing if we have a zero. I don't think this can be done at the thread level in cuda. I guess the options are:

- leave things as-is and loudly document that integer floor division by zero is undefined in cuDF. 
- leave things as-is, and raise a warning for _all_ floordiv operations involving an integer that warns that if there are zeroes, the result could be corrupt (but don't check if there actually are zeroes)
- check for zeroes and warn
- check for zeroes and raise (current approach)

The only other thing I think is worth noting is that I think the check/synchronization is already happening. There's a line in the existing code that says `(0.0 in tmp)`, the only difference is instead of raising it casts the output to `float64` so that it can then return `NaN` for those cases.

Reply:
In general, we avoid introspecting device data. Unless there's a compelling reason, we should avoid it here as well. If we can easily warn for host-side integers, I'm all for that, but we should generally avoid device synchronization. :/

It would be nice if libcudf provided some type of feedback after-the-fact, so the fast path can be taken and then we could check if a div-by-zero _did occur_ (as detected by the device), but without causing any additional synchronization. I'm not aware of any such feedback mechanisms though.

Reply:
I do agree with all of that sentiment, and think we should strive to achieve the best performance possible for all of those reasons. That said, as a user I'd rather use a slower library that's mathematically consistent than a fast one that gives me the wrong answer. 

There's some discussion of this in https://github.com/rapidsai/cudf/issues/2172. In python, this isn't allowed at all:

```
>>> 5 // 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: integer division or modulo by zero
```
The conclusion in that thread seems to be that libcudf won't be changed to do this and that the consuming library should implement the guard rails. That means that even if we wanted to match the pandas behavior in this case, we'd still need to run a kernel to insert the zeros in the correct places after the fact.

Reply:
Based on the Pandas issue and discussion here, I think the general stance is that users wouldn't be happy with us just silently returning different results than Pandas, so if we wanted to do that we would need to synchronize to raise a warning or exception. Given that both numpy integer arrays and Pandas nullable integer columns return `0` when floor dividing by `0`, we should follow that behavior. Numpy raises a warning, but Pandas doesn't in this situation, so we could take the aggressive stance of following Pandas behavior here which would allow us to avoid synchronizing.

I.E. something like:
- Get mask / gather map of input denominator where values are 0
- Do floor divide via libcudf
- Set mask / gather map of output from libcudf to 0

I believe we avoid synchronizing in this path at the cost of a couple extra operations. Technically this could be implemented in a single kernel, but I highly doubt this would ever be a bottleneck anyway.

Reply:
Connected offline with @cwharris and @kkraus14 . We decided that we'll eat the perf hit from matching the pandas behavior here, and fill with zeros to avoid breakages against pandas.